dhis-2/dhis-api/src/main/java/org/hisp/dhis/setting/LazySettings.java

```diff
@@ -0,0 +1,378 @@
+  }
+
+  @Override
+  public Set<String> keys() {
+    return Set.of(keys);
+  }
+
+  @Nonnull
+  @Override
+  @SuppressWarnings("unchecked")
+  public <E extends Enum<?>> E asEnum(@Nonnull String key, @Nonnull E defaultValue) {
+    Serializable value = orDefault(key, defaultValue);
+    if (value != null && value.getClass() == defaultValue.getClass()) return (E) value;
+    return (E) asParseValue(key, defaultValue, raw -> Enum.valueOf(defaultValue.getClass(), raw));
+  }
+
+  @Nonnull
+  @Override
+  public String asString(@Nonnull String key, @Nonnull String defaultValue) {
+    if (orDefault(key, defaultValue) instanceof String value) return value;
+    // Note: index will exist as otherwise we would have exited above
+    String rawValue = rawValues[indexOf(key)];
+    return rawValue == null ? "" : rawValue;
+  }
+
+  @Nonnull
+  @Override
+  public Date asDate(@Nonnull String key, @Nonnull Date defaultValue) {
+    if (orDefault(key, defaultValue) instanceof Date value) return value;
+    return asParseValue(key, defaultValue, LazySettings::parseDate);
+  }
+
+  @Nonnull
+  @Override
+  public Locale asLocale(@Nonnull String key, @Nonnull Locale defaultValue) {
+    if (orDefault(key, defaultValue) instanceof Locale value) return value;
+    return asParseValue(key, defaultValue, LocaleUtils::toLocale);
+  }
+
+  @Override
+  public int asInt(@Nonnull String key, int defaultValue) {
+    if (orDefault(key, defaultValue) instanceof Integer value) return value;
+    return asParseValue(key, defaultValue, Integer::valueOf);
+  }
+
+  @Override
+  public double asDouble(@Nonnull String key, double defaultValue) {
+    if (orDefault(key, defaultValue) instanceof Double value) return value;
+    return asParseValue(key, defaultValue, Double::valueOf);
+  }
+
+  @Override
+  public boolean asBoolean(@Nonnull String key, boolean defaultValue) {
+    if (orDefault(key, defaultValue) instanceof Boolean value) return value;
+    return asParseValue(key, defaultValue, Boolean::valueOf);
+  }
+
+  @Override
+  public Map<String, String> toMap() {
+    Map<String, String> map = new TreeMap<>();
+    for (int i = 0; i < keys.length; i++) {
+      map.put(keys[i], rawValues[i]);
+    }
+    return map;
+  }
+
+  @Override
+  public JsonMap<JsonMixed> toJson(boolean includeConfidential) {
+    return toJson(includeConfidential, Set.of());
+  }
+
+  @Override
+  public JsonMap<JsonMixed> toJson(boolean includeConfidential, @Nonnull Set<String> filter) {
+    Set<String> includedKeys = new HashSet<>(filter);
+    if (filter.isEmpty()) {
+      includedKeys.addAll(keysWithDefaults(type));
+      includedKeys.addAll(List.of(keys));
+    }
+    if (!includeConfidential) includedKeys.removeIf(LazySettings::isConfidential);
+    return Json.object(obj -> includedKeys.forEach(key -> obj.addMember(key, asJson(key).node())))
+        .asMap(JsonMixed.class);
+  }
+
+  @Nonnull
+  private JsonValue asJson(String key) {
```

Comment:
did we always allow storing json as a system setting? This seems like something that would be more suited in the datastore.

Reply:
The `toJson` (and `asJson`) part is purely to not break our REST API where settings were exposed as JSON in some cases and so I tried to have them stay in the original JSON type. I would hope that we can simplify this at some point and just treat all of them as JSON string as well. As you say JSON values better fit the datastore and as of now there are no complex values any more in settings.